File: create-sheet.mjs
/**
 * Creates OJT Tracker Google Sheet using the service account.
 * The service account owns the sheet, so no sharing needed.
 */
import { GoogleAuth } from "google-auth-library";
import { readFileSync } from "fs";

const CREDENTIALS_PATH = "C:\\Users\\ribaldz\\Desktop\\Shadow Stack\\scripts\\gradesheet-creds.json";

async function main() {
  const creds = JSON.parse(readFileSync(CREDENTIALS_PATH, "utf-8"));
  const auth = new GoogleAuth({
    credentials: creds,
    scopes: ["https://www.googleapis.com/auth/spreadsheets"],
  });
  const client = await auth.getClient();
  // Different versions of the library have different APIs
  let token;
  if (typeof auth.getAccessToken === "function") {
    token = await auth.getAccessToken();
  } else {
    const tr = await client.getAccessToken();
    token = tr?.token || tr;
  }
  console.log("Token obtained:", token ? token.substring(0, 20) + "..." : "FAILED");
  if (!token) throw new Error("Could not get access token");

  // Create the spreadsheet
  console.log("Creating Google Sheet...");
  const createRes = await fetch("https://sheets.googleapis.com/v4/spreadsheets", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      properties: {
        title: "OJT Requirements Tracker",
      },
      sheets: [
        {
          properties: { title: "Masterlist" },
        },
      ],
    }),
  });

  if (!createRes.ok) {
    const err = await createRes.text();
    throw new Error(`Create failed: ${err}`);
  }

  const sheet = await createRes.json();
  const sheetId = sheet.spreadsheetId;
  console.log(`✅ Sheet created!`);
  console.log(`Sheet ID: ${sheetId}`);
  console.log(`URL: https://docs.google.com/spreadsheets/d/${sheetId}`);

  // Add headers
  const headers = [
    "Timestamp", "SRCode", "Name", "Section", "Semester", "Course",
    "PreOJT", "PostOJT", "VerifiedPre", "VerifiedPost",
  ];

  const headerRes = await fetch(
    `https://sheets.googleapis.com/v4/spreadsheets/${sheetId}/values/Masterlist!A1:J1?valueInputOption=USER_ENTERED`,
    {
      method: "PUT",
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        values: [headers],
      }),
    }
  );

  if (!headerRes.ok) {
    const err = await headerRes.text();
    throw new Error(`Header write failed: ${err}`);
  }

  console.log("✅ Headers written!");
  console.log(`\nUse this Sheet ID for env: ${sheetId}`);
}

main().catch(console.error);

File: create_sheet.py
"""
Creates OJT Tracker sheet tab within the existing Grade Engine spreadsheet.
The service account already has Editor access to this sheet.
"""
import json
import os
from google.oauth2 import service_account
from googleapiclient.discovery import build

CREDS_PATH = os.path.expanduser(
    "~/Desktop/Shadow Stack/scripts/gradesheet-creds.json"
)
SHEET_ID = "1TELifh_DyzEQWd1OFQjPgFe_RYzrm0n0CogkScRoRC4"

SCOPES = ["https://www.googleapis.com/auth/spreadsheets"]

HEADERS = [
    "Timestamp", "SRCode", "Name", "Section", "Semester", "Course",
    "PreOJT", "PostOJT", "VerifiedPre", "VerifiedPost",
]


def main():
    print("Setting up OJT Tracker sheet tab...")

    creds = service_account.Credentials.from_service_account_file(
        CREDS_PATH, scopes=SCOPES
    )
    service = build("sheets", "v4", credentials=creds)

    # Check if OJT tab already exists
    spreadsheet = service.spreadsheets().get(spreadsheetId=SHEET_ID).execute()
    sheet_names = [s["properties"]["title"] for s in spreadsheet.get("sheets", [])]
    print(f"Existing sheets: {sheet_names}")

    if "Masterlist" not in sheet_names:
        # Add new sheet tab
        requests = [{
            "addSheet": {
                "properties": {"title": "Masterlist"}
            }
        }]
        service.spreadsheets().batchUpdate(
            spreadsheetId=SHEET_ID, body={"requests": requests}
        ).execute()
        print("[OK] Added Masterlist sheet tab")
    else:
        print("[OK] Masterlist already exists")

    # Write headers
    body = {"values": [HEADERS]}
    result = (
        service.spreadsheets()
        .values()
        .update(
            spreadsheetId=SHEET_ID,
            range="Masterlist!A1:J1",
            valueInputOption="USER_ENTERED",
            body=body,
        )
        .execute()
    )
    print(f"[OK] Headers written! ({result.get('updatedCells')} cells)")

    print(f"\n{'='*60}")
    print(f"  Sheet tab: Masterlist")
    print(f"  Sheet ID:  {SHEET_ID}")
    print(f"  URL:       https://docs.google.com/spreadsheets/d/{SHEET_ID}")
    print(f"{'='*60}")

    # Print the env var value we'll set on Vercel
    print(f"\n  Use this for OJT_SHEET_ID env var: {SHEET_ID}")
    return SHEET_ID


if __name__ == "__main__":
    main()
